### PR TITLE
[fix] [broker] Counter of pending send messages in Replicator incorrect if schema future not complete

### PR DESCRIPTION
**Pull request by poorbarcode**
### Motivation
The Replication task execution process works like this:
- read messages from the origin cluster
- load schema
- send to the target cluster

Because the network across the cluster is unstable, the sending task may be slower, so if the number of messages being sent is too large, the replication task will be suspended. There is a counter(`pendingMessages `) in `Replicator` indicating the count of sending messages. 

<strong>(Highlight)</strong>There is a problem that causes the counter to be inaccurate when the scheme future is incomplete. If the counter is as large as the config `replicationProducerQueueSize,` the replication task will be stuck. You can reproduce by `ReplicatorTest.testReplicationWillNotStuckByIncompleteSchemaFuture`.

### Modifications
- Make that the counter is correct when exceptions occur or the `schemaFuture` is incomplete
- Same change for `ShadowReplicator`

### Documentation
- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository:
- https://github.com/poorbarcode/pulsar/pull/52

### CODE REVIEW

**Review by codelipenghui — COMMENTED** [review]
Why not only add the pending messages after `producer.sendAsync` is called?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/GeoPersistentReplicator.java [hunk 1]
```diff
@@ -152,6 +154,7 @@ protected boolean replicateEntries(List<Entry> entries) {
 
                 // Increment pending messages for messages produced locally
                 PENDING_MESSAGES_UPDATER.incrementAndGet(this);
```
[the remark this answers is not in the document]
Comment (poorbarcode) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/GeoPersistentReplicator.java]: Already fixed

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/GeoPersistentReplicator.java [hunk 2]
```diff
@@ -184,12 +183,20 @@ protected boolean replicateEntries(List<Entry> entries) {
                     msg.setSchemaInfoForReplicator(schemaFuture.get());
                     msg.getMessageBuilder().clearTxnidMostBits();
                     msg.getMessageBuilder().clearTxnidLeastBits();
+                    // Increment pending messages for messages produced locally
+                    PENDING_MESSAGES_UPDATER.incrementAndGet(this);
+                    shouldRollbackPendingSendAdd.setTrue();
                     producer.sendAsync(msg, ProducerSendCallback.create(this, entry, msg));
```
Comment (mattisonchao) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/GeoPersistentReplicator.java:189]: Why do we need `shouldRollbackPendingSendAdd`? Is `producer.sendAsync(msg, ProducerSendCallback.create(this, entry, msg))` throws exception?

Comment (poorbarcode) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/GeoPersistentReplicator.java:186]: Already removed these lines, please take a look again

**Review by codelipenghui — APPROVED** [review]

**Review by mattisonchao — APPROVED** [review]
Nice catch!

**Review by Technoboy- — APPROVED** [review]

**Review by coderzc — APPROVED** [review]

**Comment by liangyepianzhou** [comment]
The files GeoPersistentReplicator and ShadowReplicator do not exist in branch-2.10. So this modification should not be cherry-picked to branch-2.10.

**Comment by poorbarcode** [comment]
> The files GeoPersistentReplicator and ShadowReplicator do not exist in branch-2.10. So this modification should not be cherry-picked to branch-2.10.

I removed the changes of `ShadowReplicator` when I did cherry-picking, so I re-tagged it.

**Pull request merged**